runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+         if (region)
+            region->calculateFrequencyOfExecution(&loopDepth);
+         loopDepth = getLoopNestingLevel(loopDepth);
+         testLocations.push_back(std::make_pair(cursor, loopDepth)); 
+         }
+      cursor = cursor->getNextTreeTop(); 
+      }
+   if (!testLocations.empty())
+      addRecompilationTests(comp(), testLocations);
+   return 1;
+   }
+
+
+void 
+TR_JProfilingRecompLoopTest::addRecompilationTests(TR::Compilation *comp, RecompilationTestLocations &testLocations)
+   {
+   TR_PersistentProfileInfo *profileInfo = comp->getRecompilationInfo()->findOrCreateProfileInfo();
+   TR_BlockFrequencyInfo *bfi = TR_BlockFrequencyInfo::get(profileInfo);
+   TR::CFG *cfg = comp->getFlowGraph();
+   cfg->setStructure(NULL);
+
+   for (auto testLocation = testLocations.begin(); testLocation != testLocations.end(); ++testLocation)
+      {
+      TR::TreeTop *asyncCheckTreeTop = testLocation->first;
+      TR::Node *node = asyncCheckTreeTop->getNode();
+      int32_t depth = testLocation->second;
+      if (trace())
+         traceMsg(comp, "block_%d, n%dn, depth = %d\n",asyncCheckTreeTop->getEnclosingBlock()->getNumber(), asyncCheckTreeTop->getNode()->getGlobalIndex(), depth);
+      TR_ByteCodeInfo bci = asyncCheckTreeTop->getNode()->getByteCodeInfo();
+      performTransformation(comp, "%s Add recompilation test after asyncCheck node n%dn\n", optDetailString(), node->getGlobalIndex());
+      TR::Node *root = generateBlockFrequecnyCalculationNode(comp, node, bfi, bfi->getOriginalBlockNumberToGetLiveFrequency(bci,comp));
+      
+      TR::Block *originalBlock = asyncCheckTreeTop->getEnclosingBlock();
+      TR::Block *newSplittedBlock = originalBlock->split(asyncCheckTreeTop->getNextTreeTop(), cfg, true, true);
+      TR::Block *callRecompileBlock = TR::Block::createEmptyBlock(node, comp, UNKNOWN_COLD_BLOCK_COUNT);
+      callRecompileBlock->setIsCold(true);
+      
+      TR::Node *callNode = TR::Node::createWithSymRef(node, TR::icall, 2, comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jitRetranslateCallerWithPrep, false, false, true));
+      callNode->setAndIncChild(0, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateStartPCSymbolRef()));
+      callNode->setAndIncChild(1, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateCompiledMethodSymbolRef()));
+      TR::TreeTop *callTree = TR::TreeTop::create(comp, TR::Node::create(node, TR::treetop, 1, callNode));
+      callTree->getNode()->setIsProfilingCode();
+      callRecompileBlock->append(callTree);
+      cfg->addNode(callRecompileBlock);
+      
+      TR::Node *loadBaseThreshold = TR::Node::createWithSymRef(node, TR::iload, 0, comp->getSymRefTab()->createKnownStaticDataSymbolRef(&recompileThreshold, TR::Int32));
+      TR::Node *adjustThreshold = TR::Node::create(node, TR::ishl, 2, loadBaseThreshold, TR::Node::iconst(node, (depth-1)*3));
```

Comment:
Since the depth is a compile constant atm this could just be a constant no? we only need to do the shift at runtime if it is going to be dynamic?